lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -3007,9 +3007,116 @@ LogicalResult RefDefineOp::verify() {
+      // We can never read from the inputs of an object.
+      if (element.direction == Direction::In)
+        return emitError(propAssign->getLoc(),
+                         "cannot read from an object's input port");
+      // the object itself must also be a valid source.
+      src = input;
+      continue;
+    }
+
+    // Otherwise, we assume the src is some value-producing expression.
+    return success();
+  }
+}
+
+static LogicalResult checkPropAssignDest(PropAssignOp propAssign) {
+  auto dest = propAssign.getDest();
+
+  // if we are writing to a block argument directly, the block argument
+  // must be an output port.
+  if (auto blockArg = dyn_cast<BlockArgument>(dest)) {
+    auto *definingOp = dest.getParentBlock()->getParentOp();
+    if (auto moduleLike = dyn_cast<FModuleLike>(definingOp)) {
+      auto direction = moduleLike.getPortDirection(blockArg.getArgNumber());
+      if (direction == Direction::In)
+        return emitError(propAssign->getLoc(),
+                         "cannot propassign to an input port");
+    }
+    return success();
+  }
+
+  auto result = cast<OpResult>(dest);
+  auto *definingOp = result.getDefiningOp();
+
+  // If we are writing to an instance's port, the port must be an input.
+  if (auto instanceOp = dyn_cast<InstanceOp>(definingOp)) {
+    auto direction = instanceOp.getPortDirection(result.getResultNumber());
+    if (direction == Direction::Out)
+      return emitError(propAssign->getLoc(),
+                       "cannot propassign to an instance's output port");
+    return success();
+  }
+
+  // if we are writing to an object.subfield op, the base object must be a
+  // local object declaration. We do not allow writing through multiple
+  // object subfield ops (eg `propassign object.p.q.r, "foo"`).
+  if (auto objectSubfieldOp = dyn_cast<ObjectSubfieldOp>(definingOp)) {
+    auto input = objectSubfieldOp.getInput();
+
+    // Fail if we are writing to a specific field on a port. We can only
+    // write an entire object to the port, we cannot assign output objects
+    // piecemeal.
+    if (auto blockArg = dyn_cast<BlockArgument>(input))
+      return emitError(propAssign->getLoc(),
+                       "cannot propassign to a subfield of a port");
+
+    auto *inputDefiningOp = input.getDefiningOp();
+
+    // We can assign to a local object's port, if the flow is correct.
+    if (auto objectOp = dyn_cast<ObjectOp>(inputDefiningOp)) {
+      auto objectType = objectOp.getType();
+      const auto &element = objectType.getElement(objectSubfieldOp.getIndex());
+      if (element.direction != Direction::In)
+        return emitError(propAssign.getLoc(),
+                         "cannot propassign to an object's output port");
+      return success();
+    }
+  }
+
+  // Otherwise, we assume the dest is some intermediate expression which
+  // cannot be assigned.
+  return emitError(propAssign->getLoc(),
+                   "cannot propassign to an intermediate expression");
+}
```

Comment:
This is close to existing flow determination and checking code. Can existing flow determination code be improved to support properties? Then the checks are not functions ("checkSrc"/"checkDest"), but inline checks that the source/dest have the expected flow.

This should have the benefit of factoring out the common code from `checkPropAssignSrc` and `checkPropAssignDest`.

Reply:
I don't know... flow checking for properties and hardware are similar, but the code is actually different. For connecting the ports of objects, we can't just boil (fold :) ) each side down to a flow, we have to look at whether the object is an input port, output port, local object declaration, or a port of an object, before we decide if a connection is legal. 

For example:

```py
class Foo:
  input x : String
  
class Bar:
  input foo : Inst<Foo>
  propassign foo.x, "whatever"
  ```
 
What is the flow of `foo.x`? We can't read from it, it's an input port of an object. We can't write to it, it's a port on a "remote" object.

Since properties have to be assigned using a different op than normal connects, I thought it was better to put this code in a different spot, rather than mangle the existing checkFlow code to handle the rules of properties. Maybe we can add a new Inaccessible flow to support the case above? Does that sound okay? Any other ideas? I'll think about this a bit more, too. I appreciate the suggestion :D